### PR TITLE
Swift inline route registration support

### PR DESCRIPTION
**Pull request by platedbay**
Based off of the discussion from #28, I've added a wrapper class and created a new method for us to register routes in Swift. This now allows developers who are writing Swift code to perform the inline matching:

```
self.router["/say/:title/:message"] = DPLRouteHandlerBlockWrapper(routeHandlerBlock: { (deepLink) -> Void in
    if let title = deepLink.routeParameters["title"] as? String,
        message = deepLink.routeParameters["message"] as? String {
            UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: "OK").show()
    }
})
```

I would love to start a discussion with everyone as to if this is the way we'd like to handle this moving forward. It's merely my first stab at it and would love some feedback so we can get this to the right level of doneness.

### CODE REVIEW

**Comment by platedbay** [comment]
Not sure why Travis Failed: `The simulator failed to start, or the TEST_HOST application failed to run. Simulator error: Unable to run app in Simulator; will retry 2 more times. (0 ms)`.

Speaking of which, I should probably add some tests in!

**Comment by chrismaddern** [comment]
@dasmer @basememara @alcarvalho @lxcid Would love any thoughts

**Comment by alcarvalho** [comment]
I guess this is a start for better supporting Swift if you need to register any blocks. But going forward we really should rethink the API to make it more idiomatic for Swift users.

Maybe writing a wrapper around the router or even a builder for it could be better ways of dealing with Swift. Something resembling [Quick](/Quick/Quick)'s API would be nice, like:

```swift
let router = buildRouter {
  route("/say/:title/:message") { deepLink in
    if let title = deepLink.routeParameters["title"] as? String,
        message = deepLink.routeParameters["message"] as? String {
            UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: "OK").show()
    }
  }
  
  route("/product/:id", withHandler:ProductHandler.self)
}
```

But I am not even sure if that's possible. :smile:

**Comment by platedbay** [comment]
The more I think about this, the less happy I am about this "workaround" so I looked into more as to how we could support the subscripting in Swift. This is what I got:
```
self.router["hello/world"] = unsafeBitCast( { deepLink in
    if deepLink != nil
    {
        println("REALLY???")
    }
} as DPLRouteHandlerBlock, AnyObject.self)
```
And boy is it ugggglllyyyy. Let's not do this.

In the end, using `- (void)registerBlock:(DPLRouteHandlerBlock)routeHandlerBlock forRoute:(NSString *)route;` works just as well and isn't as nearly clunky. I don't think this PR is necessary and maybe we can just update the example app to use that method as the second example.

I do agree with @alcarvalho that, moving forward, there are ways around making the library in Swift a bit less clunky (but even then it's still really easy).

**Comment by basememara** [comment]
Awesome @platedbay, thx for the Swift initiatives and samples! In my opinion I think it is very verbose tho, even the implementation @alcarvalho suggested. I would love to consume it simply by doing something like this for example:
```
self.router.route("/say/:title/:message", DPLRouteHandlerBlock.self) {
  (deepLink) -> Void in
    if let title = deepLink.routeParameters["title"] as? String,
        message = deepLink.routeParameters["message"] as? String {
            UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: "OK").show()
    }
}

self.router.route("/products/:id", DPLRouteHandler.self) {
    if let storyboard = UIApplication.sharedApplication().keyWindow?.rootViewController?.storyboard {
        return storyboard.instantiateViewControllerWithIdentifier("detail") as! DPLProductDetailViewController
    }
        
    return UIViewController()
}
```

**Comment by platedbay** [comment]
Ooh I think I see how that could work. The key is passing in the type of route handling. I should be able to throw something together tonight!

**Comment by platedbay** [comment]
@basememara @alcarvalho let me know what you guys think of what I just pushed into the branch. I didn't document everything yet but it's what I was able to get working in a relatively short amount of time. The meat of it is two new methods on the `DPLDeepLinkRouter`:
```
- (void)route:(NSString *)route viewControllerBlock:(DPLTargetViewControllerBlock)viewControllerBlock;
- (void)route:(NSString *)route routeHandlerBlock:(DPLRouteHandlerBlock)routeHandlerBlock;
```
Which then allow us to do this:
```
self.router.route("beers") { [weak self] (deepLink) -> Void in
    if let navController = self?.window?.rootViewController as? UINavigationController,
        beersViewController = navController.storyboard?.instantiateViewControllerWithIdentifier("beers") as? DPLProductTableViewController {
            navController.pushViewController(beersViewController, animated: false)
    }
}

self.router.route("test") { 
    return DPLTestViewController()
}
```

**Comment by alcarvalho** [comment]
This looks great to me, @platedbay. This approach is indeed simpler then my suggestion. It's not immutable, though, which is something we should consider for Swift in the future. :sunglasses: 

But it's already very good already and by far the best way right now. Thanks for putting the effort into this! :tada:

**Comment by platedbay** [comment]
Aside from adding the `DPLTargetViewControllerBlock`, this is simply a reordering of the parameters from `- (void)registerBlock:(DPLRouteHandlerBlock)routeHandlerBlock forRoute:(NSString *)route;`. Do we think this is something that would be worth while, or should I just add a `- (void)registerTargetViewControllerBlock:(DPLTargetViewControllerBlock)targetViewControllerBlock forRoute:(NSString *)route;` to keep it inline with the existing methods?

Basically, I'm not sure if crowding the `DPLDeepLinkRouter` with "redundant" methods is worth it.

**Comment by alcarvalho** [comment]
I think it's probably better to have just one method for each thing. And since it's better for Swift to have the block as the last argument, I vote we make this breaking change. Its benefit is high enough for this decision.

**Comment by platedbay** [comment]
I guess the call comes down to what @chrismaddern and @wessmith would prefer, especially as a change like this would probably require a bigger release. We could deprecate the old register methods and tell people to use the new ones moving forward (and remove them after a while).

That said, this type of naming convention moves away from the rest of the codebase so I can definitely understand the hesitation.

**Comment by chrismaddern** [comment]
Reviving this -- literal array syntax is now supported in Swift and it's simply a type check that stops it working correctly. @wessmith correct?

Let's get this fixed and in, perhaps to a `1.3` release instead of waiting for `2.0`?

**Comment by platedbay** [comment]
Holy old PR! How's it going @chrismaddern? We've been using the default implementing just fine in our Swift 2.3 application for a while now so this need kinda dropped off for us. Is this still needed?

**Comment by wessmith** [comment]
Triage: closing in favor of #123
